### PR TITLE
fix: fail fast if remote is not reachable

### PR DESCRIPTION
**Pull request by DeepDiver1975**
## Description
Not responding remote servers can cause bad user experience.

## How Has This Been Tested?
- call both endpoints with not existing remote servers

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Review by IljaN — APPROVED** [review]

File: changelog/unreleased/41210 [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Bugfix: Fail fast on unresponsive remote servers
+
+In case a remote/federated server is not responding the share will not be added
+and any requests will fail faster.
+
+https://github.com/owncloud/core/pull/40965
```
Comment (phil-davis) [inline changelog/unreleased/41210:6]:
```suggestion
https://github.com/owncloud/core/pull/41210
```

**Review by phil-davis — COMMENTED** [review]
and code-style needs to be fixed

**Comment by phil-davis** [comment]
Various acceptance tests for federated sharing are failing with:
```
Failed step: And user "Alice" has added the public share created from server "REMOTE" using the sharing API
        Sharing::userHasAddedPublicShareCreatedByUser
        Failed to save public share.
        'Remote is unreachable'
        Failed asserting that 'error' is not equal to 'error'.
```
Example: https://drone.owncloud.com/owncloud/core/39820/96/19

Maybe the test environment is not so good, or maybe the new code is too enthusiastic about deciding that the remote is not reachable. I will have a look...

**Comment by DeepDiver1975** [comment]
@phil-davis any idea yet what is going on here? THX a lot

File: apps/files_sharing/ajax/external.php [hunk 2]
```diff
@@ -75,6 +74,13 @@
 	\OC::$server->getUserSession()->getUser()->getUID()
 );
 
+# test remote before adding the share
+$resp = $externalManager->testRemoteUrl(\OC::$server->getHTTPClientService(), $remote);
+if ($resp === false) {
+	\OCP\JSON::error(['data' => ['message' => $l->t('Remote is unreachable')]]);
+	exit();
+}
+
```
Comment (phil-davis) [inline apps/files_sharing/ajax/external.php:83]: The failing tests are all scenarios where a user tries to add a public link share.
The "federated" share is actually a public link.
That comes through this code.

Variable `$remote` at this point has the scheme `http://` on the front.
But `testRemoteUrl` expects a URL without the scheme on the front.

I hacked the code like this:
```
$x = preg_replace("(^https?://)", "", $remote);
$resp = $externalManager->testRemoteUrl(\OC::$server->getHTTPClientService(), $x);
if ($resp === false) {
	\OCP\JSON::error(['data' => ['message' => $l->t("Remote '$x' is unreachable")]]);
	exit();
}
```

Now I run a test scenario:
```
make test-acceptance-api BEHAT_FEATURE=tests/acceptance/features/apiFederationToRoot1/savePublicShare.feature:22
...
Feature: Save public shares created by oC users

  Background:                                                                            # /home/phil/git/owncloud/core/tests/acceptance/features/apiFederationToRoot1/savePublicShare.feature:4
    Given using server "LOCAL"                                                           # FeatureContext::usingServer()
    And user "Alice" has been created with default attributes and without skeleton files # FeatureContext::userHasBeenCreatedWithDefaultAttributesAndWithoutSkeletonFiles()

  Scenario: Mount public share and then delete (local server share)                               # /home/phil/git/owncloud/core/tests/acceptance/features/apiFederationToRoot1/savePublicShare.feature:22
    Given user "Brian" has been created with default attributes and without skeleton files        # FeatureContext::userHasBeenCreatedWithDefaultAttributesAndWithoutSkeletonFiles()
    And user "Alice" has created folder "/PARENT"                                                 # FeatureContext::userHasCreatedFolder()
    And user "Alice" has created a public link share with settings                                # FeatureContext::userHasCreatedAPublicLinkShareWithSettings()
      | path        | /PARENT |
      | permissions | read    |
    And user "Brian" has added the public share created from server "LOCAL" using the sharing API # FeatureContext::userHasAddedPublicShareCreatedByUser()
      Sharing::userHasAddedPublicShareCreatedByUser
      Failed to save public share.
      'Remote '172.17.0.1:8080' is unreachable'
      Failed asserting that 'error' is not equal to 'error'.
```
But it still complains that '172.17.0.1:8080` is not reachable - that is my test installation which is running fine.

I wonder what is the next piece of this puzzle?

**Comment by phil-davis** [comment]
The failing test scenarios are the ones about Brian adding a public share.
```
runsh: Total 69 scenarios (57 passed, 12 failed)
runsh: Exit code of main run: 1
runsh: Total unexpected failed scenarios throughout the test run:
apiFederationToRoot1/savePublicShare.feature:9
apiFederationToRoot1/savePublicShare.feature:22
apiFederationToRoot1/savePublicShare.feature:49
apiFederationToRoot1/savePublicShare.feature:50
apiFederationToRoot1/savePublicShare.feature:67
apiFederationToRoot1/savePublicShare.feature:68
apiFederationToRoot1/savePublicShare.feature:71
apiFederationToRoot1/savePublicShare.feature:86
apiFederationToRoot1/savePublicShare.feature:120
apiFederationToRoot1/savePublicShare.feature:121
apiFederationToRoot1/savePublicShare.feature:140
apiFederationToRoot1/savePublicShare.feature:141


runsh: Total 63 scenarios (53 passed, 10 failed)
runsh: Exit code of main run: 1
runsh: Total unexpected failed scenarios throughout the test run:
apiFederationToShares1/savePublicShare.feature:10
apiFederationToShares1/savePublicShare.feature:23
apiFederationToShares1/savePublicShare.feature:35
apiFederationToShares1/savePublicShare.feature:63
apiFederationToShares1/savePublicShare.feature:64
apiFederationToShares1/savePublicShare.feature:67
apiFederationToShares1/savePublicShare.feature:82
apiFederationToShares1/savePublicShare.feature:96
apiFederationToShares1/savePublicShare.feature:130
apiFederationToShares1/savePublicShare.feature:131
```

The usual case of users adding a federated share that is an ordinary file share on another server, that works.

**Comment by phil-davis** [comment]
If one of the calls in `testRemoteUrl` to `testUrl` throws an exception, then the try-catch block finishes at that point, and the other calls to `testUrl` do not happen.

If I separate out each `testUrl` call into its own try-catch block, then I see that every `testUrl` call is getting `ConnectException` thrown.

I can just have it do one `testUrl` call:
```
	public function testRemoteUrl(IClientService $clientService, string $remote) {
		// cut query and|or anchor part off
		$remote = \strtok($remote, '?#');
		try {
			if ($this->testUrl($clientService, 'http://' . $remote . '/status.php', true)) {
				return 'http';
			}
		} catch (ConnectException $e) {
			// noop
		}

		return false;
	}
```

And that still gets `ConnectException`

But I can manually do:
```
curl http://172.17.0.1:8080/status.php
{"installed":true,"maintenance":false,"needsDbUpgrade":false,"version":"10.14.1.0","versionstring":"10.14.1 prealpha","edition":"Enterprise","productname":"ownCloud","product":"ownCloud"}
```

So the `status.php` endpoint is actually working.

So I suppose that there is some problem with how `testUrl` is setting up to HTTP client and doing the get request.

**Comment by DeepDiver1975** [comment]
@phil-davis THX a lot for the analysis! Much appreciated!

**Comment by DeepDiver1975** [comment]
```
"reqId":"XSMwTG4Mm0JBu2okCpJC","level":3,"time":"2024-04-18T07:53:46+00:00","remoteAddr":"192.168.8.3","user":"Brian","app":"no app in context","method":"POST","url":"\/apps\/files_sharing\/external","message":"Remote server not reachable: {\"Exception\":\"GuzzleHttp\\\\Exception\\\\ConnectException\",\"Message\":\"cURL error 35: error:1408F10B:SSL routines:ssl3_get_record:wrong version number (see https:\\\/\\\/curl.haxx.se\\\/libcurl\\\/c\\\/libcurl-errors.html) for https:\\\/\\\/server\\\/ocs-provider\\\/\",\"Code\":0,\"Trace\":\"#0 \\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/Handler\\\/CurlFactory.php(158): GuzzleHttp\\\\Handler\\\\CurlFactory::createRejection()\\n#1 \\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/Handler\\\/CurlFactory.php(110): GuzzleHttp\\\\Handler\\\\CurlFactory::finishError()\\n#2 \\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/Handler\\\/CurlHandler.php(47): GuzzleHttp\\\\Handler\\\\CurlFactory::finish()\\n#3 \\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/Handler\\\/Proxy.php(28): GuzzleHttp\\\\Handler\\\\CurlHandler->__invoke()\\n#4 \\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/Handler\\\/Proxy.php(48): GuzzleHttp\\\\Handler\\\\Proxy::GuzzleHttp\\\\Handler\\\\{closure}(*** sensitive parameters replaced ***)\\n#5 \\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/PrepareBodyMiddleware.php(35): GuzzleHttp\\\\Handler\\\\Proxy::GuzzleHttp\\\\Handler\\\\{closure}(*** sensitive parameters replaced ***)\\n#6 \\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/Middleware.php(31): GuzzleHttp\\\\PrepareBodyMiddleware->__invoke()\\n#7 \\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/RedirectMiddleware.php(71): GuzzleHttp\\\\Middleware::GuzzleHttp\\\\{closure}(*** sensitive parameters replaced ***)\\n#8 \\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/Middleware.php(66): GuzzleHttp\\\\RedirectMiddleware->__invoke()\\n#9 \\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/HandlerStack.php(75): GuzzleHttp\\\\Middleware::GuzzleHttp\\\\{closure}(*** sensitive parameters replaced ***)\\n#10 \\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/Client.php(333): GuzzleHttp\\\\HandlerStack->__invoke()\\n#11 \\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/Client.php(169): GuzzleHttp\\\\Client->transfer()\\n#12 \\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/Client.php(189): GuzzleHttp\\\\Client->requestAsync()\\n#13 \\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/ClientTrait.php(44): GuzzleHttp\\\\Client->request()\\n#14 \\\/drone\\\/src\\\/lib\\\/private\\\/Http\\\/Client\\\/Client.php(184): GuzzleHttp\\\\Client->get()\\n#15 \\\/drone\\\/src\\\/apps\\\/files_sharing\\\/lib\\\/External\\\/Manager.php(553): OC\\\\Http\\\\Client\\\\Client->get()\\n#16 \\\/drone\\\/src\\\/apps\\\/files_sharing\\\/lib\\\/External\\\/Manager.php(578): OCA\\\\Files_Sharing\\\\External\\\\Manager->testUrl()\\n#17 \\\/drone\\\/src\\\/apps\\\/files_sharing\\\/ajax\\\/external.php(78): OCA\\\\Files_Sharing\\\\External\\\\Manager->testRemoteUrl()\\n#18 \\\/drone\\\/src\\\/lib\\\/private\\\/Route\\\/Route.php(155): require_once('\\\/drone\\\/src\\\/apps...')\\n#19 \\\/drone\\\/src\\\/lib\\\/private\\\/Route\\\/Router.php(344): OC\\\\Route\\\\Route->OC\\\\Route\\\\{closure}(*** sensitive parameters replaced ***)\\n#20 \\\/drone\\\/src\\\/lib\\\/kernel.php(925): OC\\\\Route\\\\Router->match()\\n#21 \\\/drone\\\/src\\\/index.php(32): OC::handleRequest()\\n#22 {main}\",\"File\":\"\\\/drone\\\/src\\\/lib\\\/composer\\\/guzzlehttp\\\/guzzle\\\/src\\\/Handler\\\/CurlFactory.php\",\"Line\":210}"}

**Comment by DeepDiver1975** [comment]
@phil-davis mind another review? THX

File: apps/files_sharing/lib/External/Manager.php [hunk 3]
```diff
@@ -538,4 +540,84 @@ private function getShares($accepted) {
...
+		return false;
+	}
+
+	private function checkConnectivity(IClientService $clientService, string $remote): bool {
+		foreach (['http://', 'https://'] as $schema) {
+			try {
+				if ($this->testUrl($clientService, $schema . $remote . '/status.php', false, true)) {
+					\OC::$server->getLogger()->error("Remote $schema$remote/status.php reachable");
+					return true;
+				}
```
Comment (phil-davis) [inline apps/files_sharing/lib/External/Manager.php:614]:
```suggestion
				if ($this->testUrl($clientService, $schema . $remote . '/status.php', false, true)) {
					return true;
				}
				\OC::$server->getLogger()->error("Remote $schema$remote/status.php not reachable");
```

If the host:port of the remote is completely down, then `testUrl` is returning false, but I get nothing in the log about it.
Maybe move the error logging to the unhappy path, and don't log when reachable.

**Review by phil-davis — COMMENTED** [review]
If the remote system is not listening on port 80, but some other port, then the checks do not work - they get directed to the "default" port 80, but need to be directed to the correct port.

See the last commit of PR #41237 - with that extra code the tests pass for me in my dev setup that has 2 servers running on localhost port 8080 and 8180 on my laptop.

File: apps/files_sharing/lib/External/Manager.php (hunk 3 above)
Comment (DeepDiver1975) [reply apps/files_sharing/lib/External/Manager.php:614]: thx for catching this :pray:

**Comment by phil-davis** [comment]
Note: for my local testing, I found that this scenario was good to run:
```
make test-acceptance-api BEHAT_FEATURE=tests/acceptance/features/apiFederationToRoot1/savePublicShare.feature:71
```
That scenario uses both local and remote (federated) servers and a public link.
Just putting this here so that I find it easily if I need it again.

**Pull request merged**